File: 20-valid-parentheses/valid-parentheses.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        for (char c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } else {
                // No opening bracket available
                if (stack.isEmpty()) {
                    return false;
                }

                char opening = stack.pop();

                if ((c == ')' && opening != '(') ||
                    (c == '}' && opening != '{') ||
                    (c == ']' && opening != '[')) {
                    return false;
                }
            }
        }

        return stack.isEmpty();
    }
}

File: 20-valid-parentheses/valid-parentheses.cpp
#include <stack>
#include <unordered_map>
#include <string>

class Solution {
public:
    bool isValid(const std::string& s) {
        std::stack<char> stack;
        std::unordered_map<char, char> brackets = {
            {')', '('},
            {'}', '{'},
            {']', '['}
        };

        for (char c : s) {
            if (brackets.count(c)) {
                // c is a closing bracket
                if (stack.empty() || stack.top() != brackets[c]) {
                    return false;
                }
                stack.pop();
            } else {
                // c is an opening bracket
                stack.push(c);
            }
        }
        return stack.empty();
    }
};